File: Day23/Day23.py
import time

def set_up(data):
    nodeA = Node(data[0])
    prevNode = nodeA
    lastNode = None
    for i in range(1,len(data)):
        newNode = Node(data[i])
        prevNode.right = newNode
        prevNode = newNode
        lastNode = newNode
    lastNode.right = nodeA
    nodeA.left = lastNode
    return nodeA

def set_up_part2(data):
    nodeA = Node(data[0])
    prevNode = nodeA
    lastNode = None
    for i in range(1,len(data)):
        newNode = Node(data[i])
        prevNode.right = newNode
        prevNode = newNode
        lastNode = newNode
    for i in range(10,1000001):
        newNode = Node(i)
        prevNode.right = newNode
        prevNode = newNode
        lastNode = newNode
    lastNode.right = nodeA
    nodeA.left = lastNode
    return nodeA

class Node:
    def __init__(self, val):
        self.val = int(val)
        self.left = None
        self.right = None
    def __repr__(self):
        return str(self.val)
    def __str__(self):
        return str(self.val)
    def display(self):
        rtnString = str(self.val)
        nxtNode = self.right
        while nxtNode.val != self.val:
            rtnString += str(nxtNode.val)
            nxtNode = nxtNode.right
        return rtnString

def move(current_cup, maxi):
    
    # Splice three cups to the right
    spliced_start = current_cup.right
    spliced_stop = current_cup.right.right.right
    spliced_values = [spliced_start.val, spliced_start.right.val, spliced_stop.val]
    
    #Connect the current cup to the cup after the splice
    spliced_stop.right.left = current_cup
    current_cup.right = spliced_stop.right
    
    
    next_cup = current_cup.right
    original_value = current_cup.val
    value = maxi
    for i in range (current_cup.val-1, current_cup.val -4, -1):
        if i not in spliced_values:
            value = i
            break
    if value <=0:
        value = maxi
        while value in spliced_values:
            value -= 1
    # maxi = current_cup.val
    while value != next_cup.val:
        # if next_cup.val == original_value:
        #     value -= 1
        #     if value <= 0:
        #         value = maxi
        # if maxi < next_cup.val:
        #     maxi = next_cup.val

        next_cup = next_cup.right
    next_cup.right.left = spliced_stop
    spliced_stop.right = next_cup.right
    next_cup.right = spliced_start
    spliced_start.left = next_cup
    
    return current_cup.right


if __name__=='__main__':
    # sample = '389125467'
    sample = '853192647'

    # part 1
    # a = set_up(sample)
    
    # for i in range(2,102):
    #     a = move(a,9)
    #     print(f'{i} : {a.display()}')
    # print(a.display())
    # while a.val != 1:
    #     a = a.right
    # print(a.display())

    # part 2
    start = time.time()
    print(f'Start: {start}')
    a = set_up_part2(sample)
    # print(a.display())
    print('Parsed input')
    currentloop = time.time()+ 30
    last_i = 2
    for i in range(2,10000002):
        a = move(a, 100000)
        if time.time() > currentloop:
            loop_time = time.time() - (currentloop - 30)
            remaining_seconds = ((10000002 - i)/ (i-last_i)) * loop_time
            print(f'{i}/10000002 in {time.time()- start} seconds. {round(loop_time)} for {i - last_i } values. Estimate {round(remaining_seconds)/60} minutes remaining')
            last_i = i
            currentloop = time.time()+ 30
        # print(f'{i} : {a.display()}')
    # print(a.display())
    while a.val != 1:
        a = a.right
    print(a.display())

File: Day23/Day23.cpp
#include <iostream>
#include <algorithm> // for std::find
#include <iterator> // for std::begin, std::end
// #include <fstream>
// #include <string>
// #include <vector>

struct node 
{
    int data;
    node *next;
};

class linked_list
{
    private:
    node *head,*tail,*current_node;
    int maximum{0};
    public:
    linked_list()
    {
        head = NULL;
        tail = NULL;
    }
    void add_node(int n)
    {
        node *tmp = new node;
        tmp->data = n;
        tmp->next = NULL;

        if(head==NULL)
        {
            head = tmp;
            tail = tmp;
        }
        else
        {
            tail->next = tmp;
            tail = tail->next;
            tail->next = head;
        }
        if (tmp->data > maximum)
        {
            maximum = tmp->data;
        }
        
    }
    void insertNode(node *place, node *insert)
    {
        node *tmp = place->next;
        place->next = insert;
        insert->next->next->next = tmp;
    }
    void move()
    {
        # Splice three cups to the right
        node *spliced_start = current_node->next;
        int spliced_values {spliced_start->data,spliced_start->next->data, spliced_start}
        // spliced_stop = current_cup.right.right.right
        // spliced_values = [spliced_start.val, spliced_start.right.val, spliced_stop.val]
        // #Connect the current cup to the cup after the splice
        current_node->next = spliced_start->next->next->next;

        node *next_cup = current_node->next;
        int value {};
        for(int i=current_node->data-1; current_node->data-4;i--)
        {
            bool exists = std::find(std::begin(a), std::end(a), x) != std::end(a);
        } 
    }
    void display()
    {
        node *tmp;
        tmp = head;
        std::cout << tmp->data << std::endl;
        tmp = tmp->next;
        while (tmp->data != head->data)
        {
            std::cout << tmp->data << std::endl;
            tmp = tmp->next;
        }
    }
};

int main() 
{
    int input[] {8,5,3,1,9,2,6,4,7};
    linked_list a;
    for(int i : input){
        a.add_node(i);
    }
    // a.add_node(1);
    // a.add_node(2);
    a.display();
    return 0;
}